More bug fixes, improvements, and features

PaulSD opened the pull request:
Primary changes:
* Use strict mode in eval() to keep the global javascript variable namespace clean.
* Switch from `${ }` to `<$ $>` template delimiters.  (Old delimiters are still supported in cases where they were originally supported, but new delimiters are needed for some new use cases.  See #166 for details.)
* Allow template delimiters in variables.
* Allow disabling template parsing per value using a `$!` prefix.
* Add support for "static" variables that are evaluated once when the card is loaded, rather than on every update.
* Add `config` and `output` template variables.
* Add support for templates that return Promises / use asynchronous functions.  (Based on #149, with some changes/improvements.)

After this is merged, my opinion is that the dev branch is good enough to be merged into master.
However, of course, feel free to leave this in dev for now and do another beta release if you aren't comfortable with merging into master yet.

Closes #149, closes #160, closes #166.

PaulSD committed:
Tweak workflows to fix a few minor issues

PaulSD committed:
Rearrange some functions for clarity

PaulSD committed:
Use strict mode in eval() and eliminate some redundant operations

PaulSD committed:
Adjust variables to support the same template syntax as other sections

Specifically, support ${} in variables, multiple templates in variables,
and nested templates in data structures.

PaulSD committed:
Allow disabling template parsing per-value using $! prefix

PaulSD committed:
Fix global/local vars mixup

PaulSD committed:
Add support for "static" variables

These will be evaluated only on the first update/render, and will be
preserved without re-evaluation for subsequent updates.

These may be used to manage the performance impact of expensive
operations, by eliminating unnecessary re-evaluations.

PaulSD committed:
Fix bug in the parsing of multiple templates

(Was only parsing the first template, not all of them)

PaulSD committed:
Cleanup, document initialization, and prep for async changes

PaulSD committed:
Add support for async templates

Co-authored-by: Paul Donohue <git@PaulSD.com>
Co-authored-by: Mike A. <git@mikealmel.ooo>

ildar170975 commented:
Seems to be very massive work.
Separate thanks for "Upload Artifacts".
Will test.

Meanwhile, some FRs (could be implemented separately if possible):
1. Is it possible to add a `config` variable (stands for a config of a card) as available in templates?
2. Is it possible to make this scenario
```
variables:
  xxx: !secret my_super_secret
```
supported?

PaulSD commented:
> Is it possible to add a config variable (stands for a config of a card) as available in templates?

You mean you want the full original config of CTC (before evaluation of templates) available as a variable?  Yeah, I can probably do that.  Do you have an example of how you might use that, so I can make sure I implement this in a way that supports what you are trying to do?

> !secret my_super_secret

Unfortunately this one isn't easy.  `!secret` is processed by the YAML parser on the HA server, and I think it is only processed when parsing configuration.yaml (not when parsing any other YAML) ... If it isn't already being processed by HA before handing the config off to CTC, then I think the only way to handle this would be to modify HA itself to do that.  (There is nothing CTC can do using the current HA APIs to handle that.)

ildar170975 commented:
> You mean you want the full original config of CTC (before evaluation of templates) available as a variable? Yeah, I can probably do that. Do you have an example of how you might use that, so I can make sure I implement this in a way that supports what you are trying to do?

Thanks in advance! I decided to create a separate FR (https://github.com/iantrich/config-template-card/issues/160) to avoid cluttering of this PR.

> !secret is processed by the YAML parser on the HA server, and I think it is only processed when parsing configuration.yaml (not when parsing any other YAML) 

Sad... We can use secrets in lovelace yaml like:
```
type: history-graph
hours_to_show: !secret xxxxxx
```
so I thought may be it could be possible to use them inside variables.

Thanks again!
Will keep testing these changes.

PaulSD committed:
Update dependency versions

PaulSD committed:
Add `config` and `output` template variables

Closes #160

ildar170975 commented:
@PaulSD 
Analysing these changes, need some more time...
Started with Readme, you added very good clarifications there!

Note: whenever I say "I will think about a commentary", this means either "I may suggest you exact commentaries & you will add them in this PR" or "I may suggest exact commentaries in a separate PR", we will discuss what is better.

> Based on https://github.com/iantrich/config-template-card/pull/149, with some changes/improvements.)

Shall we close this PR? @malmeloo
Does the current PR cover functionality, proposed in https://github.com/iantrich/config-template-card/pull/149 ?
Update: missed your note "Closes https://github.com/iantrich/config-template-card/pull/149."

ildar170975 commented on README.md:
```diff
@@ -194,22 +236,37 @@ title: My dashboard
+Config values that begin with `${` and end with `}` are parsed as a single template, including any nested `${` and `}` sequences/characters.  For example, `${vars[0]}-${vars[1]}` will attempt to evaluate `vars[0]}-${vars[1]` and will fail due to the invalid `}` and `${`.
 
-[Troubleshooting](https://github.com/thomasloven/hass-config/wiki/Lovelace-Plugins)
+Config values that do not begin with `${` and end with `}` may contain multiple templates, however those templates cannot contain `}` characters.  For example, `${vars[0]}-${vars[1]}:` will work as expected, but `${() => { return 0; }}:` will fail due to the `}` character in the template.
```
Could you clarify why a "colon" is here?
`${vars[0]}-${vars[1]}:`
`${() => { return 0; }}:`

PaulSD replied:
To make this a valid example, I needed to have some character either before the `${` or after the `}`
A common use case might be to use a template to create a dynamic label ending with a `:` so I figured that might be a logical character to use after `}` in the example.

ildar170975 replied:
Sorry, I am not getting.
Consider this code:
```
some_option: ${vars[0]}-${vars[1]}
```
Is it considered as valid?
It contains a proper number of "opening" & "closing" brackets, so for a user it seems to be correct.

ildar170975 replied:
Noticed that adding any character after the last "}" is a way to use several templates: https://github.com/iantrich/config-template-card/issues/166#issuecomment-2708450827

PaulSD replied:
Let's continue this discussion in #166

ildar170975 commented on README.md:
```diff
@@ -39,24 +49,27 @@ resources:
+| staticVariables | list   | **Optional** | List of variables, which can be templates, that can be used in your `config` and indexed using `svars` or by name. These are evaluated only on the first update/render and are preserved without re-evaluation for subsequent updates. |
+| card            | object | **Optional** | Card configuration. (A card, row, or element configuration must be provided.) |
+| row             | object | **Optional** | Row configuration. (A card, row, or element configuration must be provided.) |
+| element         | object | **Optional** | Element configuration. (A card, row, or element configuration must be provided.) |
```
Perhaps we need to add some clarifications for beginners what an "element" is; same for the "row".
Also (unrelated): we need to add a note about "avoid placing a whole Picture elements card into CTC if templates are needed for some elements only". Have to explain it again & again in Community. Will think about a proper commentary. Similar note - about "do not use JS templates of CTC where jinja templates of card-mod can be used".

ildar170975 replied:
Proposed these clarifications in https://github.com/iantrich/config-template-card/pull/164

PaulSD replied:
Regarding JS vs Jinja templates, I'm hoping that my new explanation of those near the top of the README will help.

ildar170975 replied:
I have seen these explanations. You probably meant my remark about card-mod - it is mainly about "optimizing a code" than about "js vs jinja".

ildar170975 replied:
Let's consider this as "resolved". Will discuss additional clarifications in https://github.com/iantrich/config-template-card/pull/164

ildar170975 commented on README.md:
```diff
@@ -194,22 +236,37 @@ title: My dashboard
+
+Changes to these variable definitions are not automatically propagated to cards that are already rendered on the dashboard, so you must reload your browser after making changes.
+
+### Note: All templates must be enclosed by `${}`, except when defining variables.
```
In Readme there are examples with
```
xxx: '${}'
```
or
```
xxx: "${}"
```
or 
```
xxx: ${}
```
We should think about using ONE syntax to avoid confusion  (even if variants with `'` & `"` do work as well).
BTW, to avoid confusion (+ because of long templates), I am using [yaml-multiline syntax](https://yaml-multiline.info/) like
```
xxx: >-
  ${ ... my massive template ... }
```
Perhaps we should add a few words about it. I will think about a commentary.

PaulSD replied:
Good idea.  It is probably worth documenting that all of these are supported, but I agree that providing recommendations is a good idea.

For short templates, I typically use:
```
xxx: '${...}'
```
However, in the corner case where a short template contains single quotes and does not contain double quotes or escape characters (like `\n`), it may be more readable to surround the template with double quotes instead of single quotes (to avoid the need to escape single quotes within the template).
For long templates, I use multiline syntax as you showed above.

Except in the corner case above, double quotes should probably be avoided because the YAML parser will parse escape characters (such as `\n`) in the template, which could alter the Javascript semantics in unexpected ways.
Unquoted templates should probably also be avoided since YAML has a number of unintuitive character restrictions on those as well.  For example, the `: ` (colon space) character combination is [not supported](https://yaml.org/spec/1.2.2/#733-plain-style) in unquoted values.

PaulSD replied:
I have added documentation/conventions for this to the `Configuration` section in `README.md`, but I have not yet updated the existing `Examples`.

ildar170975 commented on README.md:
```diff
@@ -194,22 +236,37 @@ title: My dashboard
+Config values that do not begin with `${` and end with `}` may contain multiple templates, however those templates cannot contain `}` characters.  For example, `${vars[0]}-${vars[1]}:` will work as expected, but `${() => { return 0; }}:` will fail due to the `}` character in the template.
+
+Values that begin with `$!` will not be parsed for templates.  `$!` will be stripped from the beginning of the value, but any `${}` sequences within the value will be left as-is.
+
```
Some clear examples for this should be added...

ildar170975 commented on README.md:
```diff
@@ -194,22 +236,37 @@ title: My dashboard
-Both arrays and objects are supported, just like in card's local variables. It is allowed to mix the two types, i.e. use an array in dashboard variables and an object in card variables, or the other way around. If both definitions are arrays, then dashboard variables are put first in `vars`. In the mixed mode, `vars` have array indices and as well as variable names.
+Both arrays and objects are supported, just like in card's local variables. It is allowed to mix the two types, i.e. use an array in dashboard variables and an object in card variables, or the other way around. If both definitions are arrays, then dashboard variables are put first in `vars`/`svars`. In the mixed mode, `vars`/`svars` have array indices and as well as variable names.
+
+Changes to these variable definitions are not automatically propagated to cards that are already rendered on the dashboard, so you must reload your browser after making changes.
```
Could you clarify this sentence?

PaulSD replied:
I've reworded it to try to make it clearer.  Is the new version better?

ildar170975 replied:
Much better! thank you.

ildar170975 commented on README.md:
```diff
@@ -39,24 +49,27 @@ resources:
+| --------------- | ------ | ------------ | ------------------------------------------------------ |
+| type            | string | **Required** | `custom:config-template-card`                          |
+| entities        | list   | **Required** | List of entity strings that should be watched for updates. Templates can be used here. |
+| variables       | list   | **Optional** | List of variables, which can be templates, that can be used in your `config` and indexed using `vars` or by name. These are evaluated on each update/render. |
```
We need to clarify what makes the card to be "updated/rendered". Perhaps some words about "if entity declared in `entities` was changed" & "if templates depend on time".

ildar170975 replied:
Also, we need to clarify in some chapter that variable A may depend on a variable B (+ some words about an order: do not use variable B before it is declared). Same for staticVariables.

ildar170975 replied:
~Please remove a trailing point after "These are evaluated on each update/render"~

PaulSD replied:
> We need to clarify what makes the card to be "updated/rendered". Perhaps some words about "if entity declared in `entities` was changed" & "if templates depend on time".

If the CTC config is updated (using the dashboard editor), if any entity listed in `entities` was changed, or if HA sends any change event that CTC does not recognize/expect.

ildar170975 replied:
A note like this should be added nearby:
"An update occurs in following cases:
-- a config of `config-template-card` or an inner card is changed by a user;
-- any entity listed in `entities` is updated."

As for "HA sends any change event that CTC does not recognize/expect" - perhaps we need to clarify this for an average user. Or just not write about it )))

PaulSD replied:
> Also, we need to clarify in some chapter that variable A may depend on a variable B (+ some words about an order: do not use variable B before it is declared). Same for staticVariables.

Yes.  Variables may reference other previously defined variables, but cannot reference subsequently defined variables.

ildar170975 replied:
OK, I will add clarifications in https://github.com/iantrich/config-template-card/pull/164 - both for "what is an update" & "variables may reference other variables".
Let's consider this as "resolved".

ildar170975 commented on README.md:
```diff
@@ -39,24 +49,27 @@ resources:
+| type            | string | **Required** | `custom:config-template-card`                          |
+| entities        | list   | **Required** | List of entity strings that should be watched for updates. Templates can be used here. |
+| variables       | list   | **Optional** | List of variables, which can be templates, that can be used in your `config` and indexed using `vars` or by name. These are evaluated on each update/render. |
+| staticVariables | list   | **Optional** | List of variables, which can be templates, that can be used in your `config` and indexed using `svars` or by name. These are evaluated only on the first update/render and are preserved without re-evaluation for subsequent updates. |
```
We need to clarify what is a "first update". Is it "when a view containing a card was first opened after HA bootup"?

ildar170975 replied:
~Please remove a trailing point after "subsequent updates"~

PaulSD replied:
> We need to clarify what is a "first update". Is it "when a view containing a card was first opened after HA bootup"?

When a client (browser or app) loads the HA web interface.  For a browser, refreshing the page or closing/reopening the tab/window/browser causes a new "first update".  For an app, the "first update" typically only happens when the app is first opened after booting the device, or if HA is upgraded and the user clicks the "Reload UI" button, or if the user reloads the UI through the settings.

Keep in mind that all of the CTC logic/updates/rendering runs on the client, not on the server.

ildar170975 replied:
For browser: for storage-mode dashboards a refresh done by smth like F5 (or reopening); for yaml-mode dashboards - there is also a "Refresh" menu action which "rebuilds" a view (since F5 does not "rebuild" it). Will that "1st update" be performed in all these cases?
For Companion App: similar cases (swipe down instead of F5), similar question.

PaulSD replied:
This is tied to the lifecycle of Javascript objects on the client.  Each time the CTC Javascript object instance is wiped and reinitialized by the client then a new "first update" happens.

In the browser, the relevant objects are stored in page context, not in localStorage/sessionStorage, so F5 should always wipe and reinitialize CTC (resulting in a new "first update"), but clicking around HA UI elements (like switching dashboards) typically will not wipe and reinitialize the CTC object.  This is the case regardless of whether you are using storage-mode or yaml-mode dashboards.  I'm not sure whether the yaml-mode "Refresh" button reloads the page or otherwise wipes and reinitializes the CTC object - It may or may not (I haven't tested it).

I'm not really sure about the app (I also haven't tested it extensively), but I suspect swiping down, or closing and re-opening the app should wipe and reinitialize the CTC object.

ildar170975 commented on README.md:
```diff
@@ -158,34 +172,62 @@ variables:
+  type: entities
+  entities:
+    - entity: light.bed_light
+      name: "${(async () => states['light.bed_light'].state === 'on' ? 'Bed Light On' : 'Bed Light Off' )();}"
```
We need a "user-friendly" explanation what is a difference between this code and "traditional"
`name: "${ states['light.bed_light'].state === 'on' ? 'Bed Light On' : 'Bed Light Off' }"`
OR - we need to add a "preamble" like "for advanced use ...", what do you think? (not to scare people off)

PaulSD replied:
Yes, this is for advanced use.  This particular example is rather pointless/contrived - there is no functional difference between this code and "traditional" code, except that this version happens to run asynchronously for no good reason.  It is assumed that a user who needs asynchronous code would understand that this is just a trivial example to demonstrate the syntax or test asynchronous functionality.

ildar170975 replied:
I see. We need to explain things for an "average John Smith" who may not be a JS expert.
Do you have a good PRACTICAL example - and short & easy same time? 
Perhaps we need to use it...

Update: from other side, most examples here are abstract & have no practical use; so, imho an example for "async" could be not so "practical". Anyway, let's consider some more interesting examples if possible.

PaulSD replied:
The primary use case for this would probably be AJAX calls to an API to retrieve data.  I can probably come up with a basic example, but let's take care of some of these other issues first, then circle back to this later.

ildar170975 commented on README.md:
```diff
@@ -158,34 +172,62 @@ variables:
-## Dashboard wide variables
+When defining `entities` that use templates, any `entities` templates that use asynchronous functions will be skipped, and any referenced `vars` that use asynchronous functions will contain the unsettled `Promise` object, which will not settle before `entities` is used. However, any referenced `svars` that use asynchronous functions will have complete/settled values.
+
+(The reason that asynchronous functions cannot be used for `entities` is that Lit does not support asynchronous functions in `shouldUpdate()` where `entities` is evaluated and used.  The alternative [lovelace-card-templater](https://github.com/gadgetchnnel/lovelace-card-templater) cannot support templates in its `entities` for the same reason; It uses API calls to render templates, and API calls  ...
```
typo: `it uses API calls` (not capital)

PaulSD replied:
Fixed.

ildar170975 commented on README.md:
```diff
@@ -158,34 +172,62 @@ variables:
+
+When defining `staticVariables` that reference other (previously defined) `svars`, any referenced `svars` that use asynchronous functions will contain the unsettled `Promise` object.
+
+Similarly, when defining `variables` that reference other (previously defined) `vars`, any referenced `vars` that use asynchronous functions will contain the unsettled `Promise` object. However, any referenced `svars` that use asynchronous functions will contain complete/settled values.
```
> However, any referenced `svars` that use asynchronous functions will contain complete/settled values.

Is there any chance that `svars[i]` is not resolved yet to a settled value by this moment?

PaulSD replied:
No.  At the time the `variables` are evaluated, all staticVariables/svars are guaranteed to have settled.  Or, to put it another way, `variables` is not evaluated until after all staticVariables/svars have settled.

ildar170975 replied:
Good. The "At the time the variables are evaluated, all staticVariables/svars are guaranteed to have settled" should be added here somehow, it says important things with simple words.

ildar170975 commented on README.md:
```diff
@@ -158,34 +172,62 @@ variables:
+Similarly, when defining `variables` that reference other (previously defined) `vars`, any referenced `vars` that use asynchronous functions will contain the unsettled `Promise` object. However, any referenced `svars` that use asynchronous functions will contain complete/settled values.
 
-## Dashboard wide variables
+When defining `entities` that use templates, any `entities` templates that use asynchronous functions will be skipped, and any referenced `vars` that use asynchronous functions will contain the unsettled `Promise` object, which will not settle before `entities` is used. However, any referenced `svars` that use asynchronous functions will have complete/settled values.
```
> However, any referenced `svars` that use asynchronous functions will contain complete/settled values.

Same remark as above.

ildar170975 commented on README.md:
```diff
@@ -39,24 +49,27 @@ resources:
+| Name            | Type   | Requirement  | Description                                            |
+| --------------- | ------ | ------------ | ------------------------------------------------------ |
+| type            | string | **Required** | `custom:config-template-card`                          |
+| entities        | list   | **Required** | List of entity strings that should be watched for updates. Templates can be used here. |
```
Please remove a trailing point

PaulSD replied:
Since most of the descriptions included multiple sentences, it seems to me that all of the sentences (including the last sentence in each description) should end with a period.  The only one that doesn't end with a period is the description for "type", and I left that one off because it must be an exact value that must not end with a period.

ildar170975 replied:
Not crucial, just personal styling habits.

ildar170975 commented on README.md:
```diff
@@ -39,24 +49,27 @@ resources:
+| `states`    | The [states](https://developers.home-assistant.io/docs/frontend/data/#hassstates) object |
+| `user`      | The [user](https://developers.home-assistant.io/docs/frontend/data/#hassuser) object |
 | `vars`      | Defined by `variables` configuration and accessible in your templates to help clean them up. If `variables` in the configuration is a yaml list, then `vars` is an array starting at the 0th index as your firstly defined variable. If `variables` is an object in the configuration, then `vars` is a string-indexed map and you can also access the variables by name without using `vars` a ...
+| `svars`     | Defined by `staticVariables` configuration and accessible in your templates to avoid redundant expensive operations. If `staticVariables` in the configuration is a yaml list, then `svars` is an array starting at the 0th index as your firstly defined variable. If `staticVariables` is an object in the configuration, then `svars` is a string-indexed map and you can also access the va ...
```
~Please remove a trailing point after "at all"~

ildar170975 commented:
Related to "Markdown card example" added in 153: please correct indentation for a `card` part.
Update: did it myself in https://github.com/iantrich/config-template-card/pull/165 (it also has a typo in a resource key)

ildar170975 commented on README.md:
```diff
@@ -69,8 +69,10 @@ Exactly one of `card`, `row`, or `element` is required.
 | `hass`      | The [hass](https://developers.home-assistant.io/docs/frontend/data/) object.       |
 | `states`    | The [states](https://developers.home-assistant.io/docs/frontend/data/#hassstates) object. |
 | `user`      | The [user](https://developers.home-assistant.io/docs/frontend/data/#hassuser) object. |
+| `config`    | The configuration provided to config-template-card. This is read-only, includes the unevaluated templates, and does not include any dashboard-wide `config_template_card_staticVars` or `config_template_card_vars` values. |
```
We need to add some examples for this option. I will think about good use-cases. Not sure what to do with "templated options" though...

PaulSD replied:
The assumption here is that you only need to read non-template options.  If you need to read template options then you would probably use `output` instead.

ildar170975 replied:
I have not tested with `output` yet. Does it contain "unresolved" code? Or "resolved"?

ildar170975 replied:
BTW - please check https://github.com/iantrich/config-template-card/issues/161#issuecomment-2708511970, a good practical example for `config`.

ildar170975 replied:
For `config`: added a simple example in https://github.com/iantrich/config-template-card/pull/169

ildar170975 commented on README.md:
```diff
@@ -69,8 +69,10 @@ Exactly one of `card`, `row`, or `element` is required.
+| `config`    | The configuration provided to config-template-card. This is read-only, includes the unevaluated templates, and does not include any dashboard-wide `config_template_card_staticVars` or `config_template_card_vars` values. |
 | `svars`     | Defined by `staticVariables` configuration and accessible in your templates to avoid redundant expensive operations. If `staticVariables` in the configuration is a yaml list, then `svars` is an array starting with index 0. If `staticVariables` in the configuration is an object, then `svars` is a string-indexed map and you can also access the variables by name without using `svars` ...
 | `vars`      | Defined by `variables` configuration and accessible in your templates to help clean them up. If `variables` in the configuration is a yaml list, then `vars` is an array starting with index 0. If `variables` in the configuration is an object, then `vars` is a string-indexed map and you can also access the variables by name without using `vars` at all. |
+| `output`    | While evaluating the `card`, `row`, or `element` configuration, `output` contains the partially assembled/evaluated copy of that config section. This may be used, for example, to reference a nested card's `type` or `entities` in a template for that card's `name` without requiring the referenced values to be defined in `variables`. |
```
Could you provide an example for using this option?

PaulSD committed:
README.md: Move the template delimiter docs closer to the top

PaulSD committed:
Switch multi-template syntax to use new template delimiter `<$ $>`

The original `${ }` syntax is still supported for single-template
values, but it caused too many problems for multi-template values.  See

Closes #166

PaulSD committed:
README.md: Misc updates/additions

PaulSD commented:
(The last commit at the moment is docs I am still working on.)

PaulSD commented:
After rethinking the delimiters, I'm also rethinking a few other aspects of these changes ... Marking this PR as a draft until I finish refactoring.
